java-api-core/src/com/kinvey/java/store/RealtimeRouter.java

```diff
@@ -0,0 +1,99 @@
+package com.kinvey.java.store;
+
+import com.kinvey.java.AbstractClient;
+import com.pubnub.api.PNConfiguration;
+import com.pubnub.api.PubNub;
+import com.pubnub.api.callbacks.SubscribeCallback;
+import com.pubnub.api.models.consumer.PNStatus;
+import com.pubnub.api.models.consumer.pubsub.PNMessageResult;
+import com.pubnub.api.models.consumer.pubsub.PNPresenceEventResult;
+
+import java.util.Collections;
+
+/**
+ * Created by yuliya on 2/15/17.
+ */
+
+class RealtimeRouter {
+
+    private static final Object lock = new Object();
+    private static RealtimeRouter realtimeRouter;
+    private PubNub pubnubClient;
+    private String channelGroup;
+    private AbstractClient client;
+    private SubscribeCallback subscribeCallback;
+
+    private RealtimeRouter() {
+
+    }
+
+    static RealtimeRouter getInstance() {
+        synchronized (lock) {
+            if (realtimeRouter == null) {
+                realtimeRouter = new RealtimeRouter();
+            }
+            return realtimeRouter;
+        }
+    }
+
+    void initialize(String channelGroup, String publishKey, String subscribeKey, String authKey, AbstractClient client) {
+        synchronized (lock) {
+            if (!isInitialized()) {
```

Comment:
Do we need to acquire this lock every time initialized is called, even if the `RealtimeRouter` has already been initialized? We may be able to make this more efficient by doing [double-check locking](https://en.wikipedia.org/wiki/Double-checked_locking).

Reply:
I agree. Updated.